File: src/main/java/org/jcodec/containers/mp4/boxes/Box.java
package org.jcodec.containers.mp4.boxes;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;

import org.jcodec.common.StringUtils;
import org.jcodec.common.UsedViaReflection;
import org.jcodec.common.io.NIOUtils;
import org.jcodec.common.tools.ToJSON;
import org.jcodec.containers.mp4.IBoxFactory;
import org.jcodec.platform.Platform;

import static org.jcodec.common.Preconditions.checkState;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * An MP4 file struncture (box).
 * 
 * @author The JCodec project
 * 
 */
public abstract class Box {
    public Header header;
    public static final int MAX_BOX_SIZE = 128 * 1024 * 1024;
    
    @Retention(RetentionPolicy.RUNTIME)
    @Target(ElementType.METHOD)
    public static @interface AtomField {
        int idx();
    }
    
    @UsedViaReflection
    public Box(Header header) {
        this.header = header;
    }

    public Header getHeader() {
        return header;
    }

    public abstract void parse(ByteBuffer buf);

    public void write(ByteBuffer buf) {
        ByteBuffer dup = buf.duplicate();
        NIOUtils.skip(buf, 8);
        doWrite(buf);

        header.setBodySize(buf.position() - dup.position() - 8);
        checkState(header.headerSize() == (long) 8);
        header.write(dup);
    }

    protected abstract void doWrite(ByteBuffer out);
    
    public abstract int estimateSize();

    public String getFourcc() {
        return header.getFourcc();
    }

    public String toString() {
        StringBuilder sb = new StringBuilder();
        dump(sb);
        return sb.toString();

    }

    protected void dump(StringBuilder sb) {
        sb.append("{\"tag\":\"" + header.getFourcc() + "\"}");
    }

    public static Box terminatorAtom() {
        return createLeafBox(new Header(Platform.stringFromBytes(new byte[4])), ByteBuffer.allocate(0));
    }

    public static String[] path(String path) {
        return StringUtils.splitC(path, '.');
    }

    public static LeafBox createLeafBox(Header atom, ByteBuffer data) {
        LeafBox leaf = new LeafBox(atom);
        leaf.data = data;
        return leaf;
    }

    public static Box parseBox(ByteBuffer input, Header childAtom, IBoxFactory factory) {
        Box box = factory.newBox(childAtom);
    
        if (childAtom.getBodySize() < Box.MAX_BOX_SIZE) {
            box.parse(input);
            return box;
        } else {
            return new LeafBox(Header.createHeader("free", 8));
        }
    }

    public static <T extends Box> T asBox(Class<T> class1, Box box) {
        try {
            T res = Platform.newInstance(class1, new Object[]{box.getHeader()});
            ByteBuffer buffer = ByteBuffer.allocate((int)box.getHeader().getBodySize());
            box.doWrite(buffer);
            ((java.nio.Buffer)buffer).flip();
            res.parse(buffer);
            return res;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
    
    public static class LeafBox extends Box {
        ByteBuffer data;

        public LeafBox(Header atom) {
            super(atom);
        }
        
        public LeafBox(Header atom, ByteBuffer data) {
            super(atom);
            this.data = data;
        }

        public void parse(ByteBuffer input) {
            data = NIOUtils.read(input, (int) header.getBodySize());
        }

        public ByteBuffer getData() {
            return data.duplicate();
        }

        @Override
        protected void doWrite(ByteBuffer out) {
            NIOUtils.write(out, data);
        }

        @Override
        public int estimateSize() {
            return data.remaining() + Header.estimateHeaderSize(data.remaining());
        }
    }
    
}

File: src/main/java/org/jcodec/containers/mp4/boxes/Box.kt
package org.jcodec.containers.mp4.boxes

import org.jcodec.common.Preconditions
import org.jcodec.common.StringUtils
import org.jcodec.common.UsedViaReflection
import org.jcodec.common.io.NIOUtils
import org.jcodec.containers.mp4.IBoxFactory
import org.jcodec.platform.Platform
import java.nio.ByteBuffer

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * An MP4 file struncture (box).
 *
 * @author The JCodec project
 */
abstract class Box @UsedViaReflection constructor(var header: Header) {

    abstract fun parse(buf: ByteBuffer)
    fun write(buf: ByteBuffer) {
        val dup = buf.duplicate()
        NIOUtils.skip(buf, 8)
        doWrite(buf)
        header.setBodySize(buf.position() - dup.position() - 8)
        Preconditions.checkState(header.headerSize() == 8.toLong())
        header.write(dup)
    }

    protected abstract fun doWrite(out: ByteBuffer)
    abstract fun estimateSize(): Int
    val fourcc: String
        get() = header.fourcc!!

    override fun toString(): String {
        val sb = StringBuilder()
        dump(sb)
        return sb.toString()
    }

    open fun dump(sb: StringBuilder) {
        sb.append("{\"tag\":\"" + header.fourcc + "\"}")
    }

    class LeafBox(atom: Header) : Box(atom) {
        var _data: ByteBuffer? = null
        override fun parse(input: ByteBuffer) {
            _data = NIOUtils.read(input, header.bodySize.toInt())
        }

        fun getData(): ByteBuffer {
            return _data!!.duplicate()
        }

        override fun doWrite(out: ByteBuffer) {
            NIOUtils.write(out, _data)
        }

        override fun estimateSize(): Int {
            return _data!!.remaining() + Header.estimateHeaderSize(_data!!.remaining())
        }
    }

    companion object {
        const val MAX_BOX_SIZE = 128 * 1024 * 1024
        fun terminatorAtom(): Box {
            return createLeafBox(Header(Platform.stringFromBytes(ByteArray(4))), ByteBuffer.allocate(0))
        }

        @JvmStatic
        fun path(path: String?): Array<String?> {
            return StringUtils.splitC(path, '.')
        }

        @JvmStatic
        fun createLeafBox(atom: Header, data: ByteBuffer?): LeafBox {
            val leaf = LeafBox(atom)
            leaf._data = data
            return leaf
        }

        @JvmStatic
        fun parseBox(input: ByteBuffer?, childAtom: Header, factory: IBoxFactory): Box {
            val box = factory.newBox(childAtom)
            return if (childAtom.bodySize < MAX_BOX_SIZE) {
                box.parse(input!!)
                box
            } else {
                LeafBox(Header.createHeader("free", 8))
            }
        }

        @JvmStatic
        fun <T : Box?> asBox(class1: Class<T>?, box: Box): T? {
            val res = Platform.newInstance(class1, arrayOf<Any>(box.header))
            return box.reinterpret(res!!) as? T?
        }
    }

    fun reinterpret(other: Box): Box? {
        return try {
            val buffer = ByteBuffer.allocate(this.header.bodySize.toInt())
            this.doWrite(buffer)
            buffer.flip()
            other.parse(buffer)
            other
        } catch (e: Exception) {
            null
        }
    }

}